File: backend-java/src/main/java/com/backend/java/backendjava/exception/ExceptionController.java
package com.backend.java.backendjava.exception;

import jakarta.persistence.EntityNotFoundException;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

@RestControllerAdvice
public class ExceptionController {
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public Map<String, String> handleValidationExceptions(MethodArgumentNotValidException ex) {
        Map<String, String> errors = new HashMap<>();
        ex.getBindingResult().getAllErrors().forEach((error) -> {
            String fieldName = ((FieldError) error).getField();
            String errorMessage = error.getDefaultMessage();
            errors.put(fieldName, errorMessage);
        });
        return errors;
    }

    @ExceptionHandler(RuntimeException.class)
    public ResponseEntity<ExceptionDetails> handleException(RuntimeException e, HttpServletRequest request) {
        ExceptionDetails exceptionDetails = new ExceptionDetails(LocalDateTime.now(), 400, e.getMessage(),
                request.getRequestURI());
        return ResponseEntity.status(400).body(exceptionDetails);
    }

    @ExceptionHandler(EntityNotFoundException.class)
    public ResponseEntity<ExceptionDetails> handleException(EntityNotFoundException e, HttpServletRequest request) {
        ExceptionDetails exceptionDetails = new ExceptionDetails(LocalDateTime.now(), 400, e.getMessage(),
                request.getRequestURI());
        return ResponseEntity.status(404).body(exceptionDetails);
    }

    @ExceptionHandler(NoSuchAlgorithmException.class)
    public ResponseEntity<ExceptionDetails> handleException(NoSuchAlgorithmException e, HttpServletRequest request) {
        ExceptionDetails exceptionDetails = new ExceptionDetails(LocalDateTime.now(), 400, e.getMessage(),
                request.getRequestURI());
        return ResponseEntity.status(400).body(exceptionDetails);
    }

}

File: backend-kotlin/src/main/kotlin/com/backend/kotlin/backendkotlin/exceptions/ExceptionController.kt
package com.backend.kotlin.backendkotlin.exceptions

import jakarta.persistence.EntityNotFoundException
import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.FieldError
import org.springframework.validation.ObjectError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.security.NoSuchAlgorithmException
import java.time.LocalDateTime
import java.util.function.Consumer


@RestControllerAdvice
class ExceptionController {
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationExceptions(ex: MethodArgumentNotValidException): Map<String, String?> {
        val errors: MutableMap<String, String?> = HashMap()
        ex.bindingResult.allErrors.forEach(Consumer { error: ObjectError ->
            val fieldName = (error as FieldError).field
            val errorMessage = error.getDefaultMessage()
            errors[fieldName] = errorMessage
        })
        return errors
    }

    @ExceptionHandler(RuntimeException::class)
    fun handleException(e: RuntimeException, request: HttpServletRequest): ResponseEntity<ExceptionDetails> {
        val exceptionDetails = ExceptionDetails(
            LocalDateTime.now(), 400,
            e.message!!,
            request.requestURI
        )
        return ResponseEntity.status(400).body(exceptionDetails)
    }

    @ExceptionHandler(EntityNotFoundException::class)
    fun handleException(e: EntityNotFoundException, request: HttpServletRequest): ResponseEntity<ExceptionDetails> {
        val exceptionDetails = ExceptionDetails(
            LocalDateTime.now(), 400,
            e.message!!,
            request.requestURI
        )
        return ResponseEntity.status(404).body(exceptionDetails)
    }

    @ExceptionHandler(NoSuchAlgorithmException::class)
    fun handleException(e: NoSuchAlgorithmException, request: HttpServletRequest): ResponseEntity<ExceptionDetails> {
        val exceptionDetails = ExceptionDetails(
            LocalDateTime.now(), 400,
            e.message!!,
            request.requestURI
        )
        return ResponseEntity.status(400).body(exceptionDetails)
    }
}